Orange/widgets/model/owneuralnetwork.py

```diff
@@ -81,6 +123,106 @@ def get_hidden_layers(self):
+        self.model = None
+        if self.check_data():
+            self.__update()
+        else:
+            self.Outputs.model.send(self.model)
+
+    @Slot(float)
+    def setProgressValue(self, value):
+        assert self.thread() is QThread.currentThread()
+        self.progressBarSet(value)
+
+    def __update(self):
+        if self._task is not None:
+            # First make sure any pending tasks are cancelled.
+            self.cancel()
+        assert self._task is None
+
+        self.setBlocking(True)
+
+        self._task = task = Task()
+
+        # A thread safe way to invoke a method
+        set_progress = methodinvoke(self, "setProgressValue", (float,))
+
+        max_iter = self.learner.kwargs["max_iter"]
+
+        def callback(iteration):
+            if task.cancelled:
+                raise CancelThreadException()  # this stop the thread
+            set_progress(iteration/max_iter*100)
+
+        # copy to set the callback so that the learner output is not modified
+        # (currently we can not pass callbacks to learners __call__)
+        learner = copy.copy(self.learner)
+        learner.callback = callback
+
+        def build_model(data, learner):
+            return learner(data)
+
+        build_model_func = partial(build_model, self.data, learner)
+
+        self.progressBarInit()
+
+        task.future = self._executor.submit(build_model_func)
+        task.watcher = FutureWatcher(task.future)
+        task.watcher.done.connect(self._task_finished)
+
+    @Slot(concurrent.futures.Future)
+    def _task_finished(self, f):
+        """
+        Parameters
+        ----------
+        f : Future
+            The future instance holding the built model
+        """
+        assert self.thread() is QThread.currentThread()
+        assert self._task is not None
+        assert self._task.future is f
+        assert f.done()
+
+        self.setBlocking(False)
+
+        self._task = None
+        self.progressBarFinished()
+
+        try:
+            self.model = f.result()
+        except Exception as ex:  # pylint: disable=broad-except
+            # Log the exception with a traceback
+            log = logging.getLogger()
+            log.exception(__name__, exc_info=True)
+            self.model = None
+            self.show_fitting_failed(ex)
+        else:
+            self.model.name = self.learner_name
+            self.model.instances = self.data
+            self.Outputs.model.send(self.model)
+
+    def cancel(self):
+        """
+        Cancel the current task (if any).
+        """
+        if self._task is not None:
+            self._task.cancel()
+            assert self._task.future.done()
+            # disconnect the `_task_finished` slot
+            self._task.watcher.done.disconnect(self._task_finished)
+            self._task = None
+        # threads use signals to run functions in the main thread and some
+        # can still be quoued (perhaps change)
+        qApp.processEvents()
```

Comment:
Remove this. If needed route the progress updates via an intermediary QObject [maybe like this](https://github.com/ales-erjavec/orange3/commit/03c6791c81302989afbfe9c046ba5dac6e41de24)

Reply:
Thanks! I haven't thought of this.